src/libfetchers/git.cc

```diff
@@ -335,104 +333,14 @@ struct GitInputScheme : InputScheme
-               way unknown errors can lead to a failure instead of
-               continuing through the wrong code path. */
-            env["LC_ALL"] = "C";
-
-            /* Check whether HEAD points to something that looks like
-               a commit, since that is the ref we want to use later
-               on. */
-            auto result = runProgram(RunOptions {
-                .program = "git",
-                .args = { "-C", repoInfo.url, "--git-dir", repoInfo.gitDir, "rev-parse", "--verify", "--no-revs", "HEAD^{commit}" },
-                .environment = env,
-                .mergeStderrToStdout = true
-            });
-            auto exitCode = WEXITSTATUS(result.first);
-            auto errorMessage = result.second;
-
-            if (errorMessage.find("fatal: not a git repository") != std::string::npos) {
-                throw Error("'%s' is not a Git repository", repoInfo.url);
-            } else if (errorMessage.find("fatal: Needed a single revision") != std::string::npos) {
-                // indicates that the repo does not have any commits
-                // we want to proceed and will consider it dirty later
-            } else if (exitCode != 0) {
-                // any other errors should lead to a failure
-                throw Error("getting the HEAD of the Git tree '%s' failed with exit code %d:\n%s", repoInfo.url, exitCode, errorMessage);
-            }
-
-            repoInfo.hasHead = exitCode == 0;
-
-            try {
-                if (repoInfo.hasHead) {
-                    // Using git diff is preferrable over lower-level operations here,
-                    // because it's conceptually simpler and we only need the exit code anyways.
-                    auto gitDiffOpts = Strings({ "-C", repoInfo.url, "--git-dir", repoInfo.gitDir, "diff", "HEAD", "--quiet"});
-                    if (!repoInfo.submodules) {
-                        // Changes in submodules should only make the tree dirty
-                        // when those submodules will be copied as well.
-                        gitDiffOpts.emplace_back("--ignore-submodules");
-                    }
-                    gitDiffOpts.emplace_back("--");
-                    runProgram("git", true, gitDiffOpts);
-
-                    repoInfo.isDirty = false;
-                }
-            } catch (ExecError & e) {
-                if (!WIFEXITED(e.status) || WEXITSTATUS(e.status) != 1) throw;
-            }
-        }
+        if (!input.getRef() && !input.getRev() && repoInfo.isLocal)
+            repoInfo.workdirInfo = GitRepo::openRepo(CanonPath(repoInfo.url))->getWorkdirInfo();
 
         return repoInfo;
     }
 
-    std::set<CanonPath> listFiles(const RepoInfo & repoInfo) const
-    {
-        auto gitOpts = Strings({ "-C", repoInfo.url, "--git-dir", repoInfo.gitDir, "ls-files", "-z" });
-        if (repoInfo.submodules)
-            gitOpts.emplace_back("--recurse-submodules");
```

Comment:
This commit changes behavior with regards to flakes residing in submodules. I wrote a regression test for that in https://github.com/NixOS/nix/pull/8064